### ISSUE TITLE
Make the "change an image" flow clearer

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#11952]
10up ran some great testing sessions where one particular hurdle was highlighted — users don't immediately pick up on the "pencil" icon as the way to edit or swap the image. More info at https://10up.com/blog/2018/testing-the-gutenberg-publishing-userflow/

This has come up anecdotally before as well.

![edit-image-button](https://user-images.githubusercontent.com/548849/48592930-758e6f80-e929-11e8-82af-a1c28a168410.png)

One possibility is to use a text button instead:

![image](https://user-images.githubusercontent.com/548849/48593105-1846ee00-e92a-11e8-9d0d-577be279c0cf.png)

Another could be to place the edit button within the image (on selection).

It'd also be nice to support "drag to swap" on the image block. This has also come up before.

**Comment by drawcard** [comment]
Dunno if text is the best solution - "Edit" seems as ambiguous as the pencil.
Perhaps a custom icon is called for, like a picture with a pencil beside it?

**Comment by melchoyce** [comment]
> Another could be to place the edit button within the image (on selection).

I'd be a little worried about having icon overload at that point.

Writing out "edit" is a nice improvement, but feels more like a bandaid. I think this is something we'll likely have to iterate on a couple times before we find something that feels right.

One note — I actually find this pattern much more jarring on other blocks, where it removes the media/embed and brings you back to the placeholder. I always expect it to open the media uploader instead of the placeholder. Whatever we do here, let's figure out a consistent approach for all blocks.

> It'd also be nice to support "drag to swap" on the image block. This has also come up before.

We might want to ask folks what they'd expect to happen if you dragged an image onto an image block. My gut reaction is "it would transform into a gallery," but that could just be me.

**Comment by mtias** [comment]
Another interaction it could support is double-click to edit / replace on the image.

**Comment by drawcard** [comment]
@mtias I am not too sure about that idea as it would be hiding functionality to the point where it's completely invisible. But it could work in conjunction with the standard button that exists at the moment.

**Comment by mtias** [comment]
Yes, it wouldn't be the main interaction, just a possible addition to it.

**Comment by drawcard** [comment]
If that's the case I'm all for that :)

**Comment by draganescu** [comment]
I'm working on a PR to test the double click as an option :)

**Comment by draganescu** [comment]
My PR made it to master, so now @mtias the image block has both a better icon to edit, a clearer flow and double click as a means to switch modes. I am continuing to add the same flow to all other similar blocks here: #14788

**Comment by draganescu** [comment]
Closing this since it moved ahead with the #14142 merged PR and now the focus is on #14795

**Comment by mtias** [comment]
Reopening because I think this is still problematic and not as clear as it could be. (Icon is still confusing.)

**Comment by draganescu** [comment]
@mtias do you have anything else in mind other than the icon? I ask because I am working here #16200 to make a component that will port this to all the media components, and it would be great to make it better while it is not ready yet :)

**Comment by mtias** [comment]
Yes, I still think text based control for this would be much clearer. Interestingly, Google Docs is doing a similar thing:
![image](https://user-images.githubusercontent.com/548849/61948392-05e87e00-afa8-11e9-9069-df6d746b3285.png)

With direct actions:
![image](https://user-images.githubusercontent.com/548849/61948444-244e7980-afa8-11e9-9a6d-98eaa89b172c.png)

**Comment by mtias** [comment]
We can try something like this:

![image](https://user-images.githubusercontent.com/548849/61999838-48a27700-b0ca-11e9-945e-cf5fe0882d27.png)

The "edit image" flow should probably be separate, specially once we surface some edit options directly (crop, flip, rotate, etc).

**Comment by mapk** [comment]
This looks great @mtias. Thanks for your mockup. Let's roll with it, @draganescu. This will essentially remove the need to return to the block placeholder screen, correct?

**Comment by mtias** [comment]
> This will essentially remove the need to return to the block placeholder screen, correct?

Yes, exactly, which was a pain to make work well in smaller containers.

**Comment by draganescu** [comment]
Ok. then I'll focus this new component to use this menu and flow instead here: #16200

**Comment by paaljoachim** [comment]
The most resent image block in 6.2 RC looks like this:

![Screen Shot 2019-07-30 at 21 43 50](https://user-images.githubusercontent.com/5323259/62159986-40894800-b313-11e9-9dd6-6ec9d8ed3f06.png)

It is a new icon that shows a kind of swap image with another. I know Kjell @kjellr worked on the icon in another issue some time ago.

**Comment by kjellr** [comment]
> To me the swap image icon seems pretty clear.

I've seen _many_ users miss this icon (in both the old and new iterations), so I'm definitely up for trying this new, clearer proposal. I also like how this would avoid the need for opening up the placeholder again. Once you have an image assigned, that interface is no longer technically a placeholder. Representing it that way has always seemed a little confusing. 

If the new flow ends up working out, I wonder if we should consider eliminating that "Go back to the placeholder to edit" flow for embeds too:

![edit](https://user-images.githubusercontent.com/1202812/62161233-abbb2600-b2e3-11e9-93da-96e00f884b91.gif)

**Comment by mapk** [comment]
I'm trying to think through the URL portion of this new UI. If I select "Insert from URL", what happens? Does the user get taken to the Media Library Modal? Should the URL input appear similar to how it displays for an Image block?

![Screen Shot 2019-07-30 at 2 38 49 PM](https://user-images.githubusercontent.com/617986/62170491-68ab8380-b2e0-11e9-8c7f-2100ff7ef597.png)

Or could it work inline like this [prototype](https://www.figma.com/file/dmv0ZfIjMxKr9YZGS5r9Gl/Replace-Image?node-id=0%3A1) below?

![replaceimage](https://user-images.githubusercontent.com/617986/62170509-719c5500-b2e0-11e9-9d58-aa559dd3381d.gif)

**Comment by paaljoachim** [comment]
Brain storming. Using the image block as an example.
The need to create an association between the existing swap image icon and the initial placeholder screen.

As seen today:
![Screen Shot 2019-07-31 at 11 35 22](https://user-images.githubusercontent.com/5323259/62201396-a1526800-b387-11e9-989f-e85841620d52.png)

Adding the swap image icon
![Swap-image-icon-image-placeholder](https://user-images.githubusercontent.com/5323259/62201815-64d33c00-b388-11e9-8c64-1cc6d1dd9077.jpg)

**Comment by shaunandrews** [comment]
Part of the confusion is related to the block icon in the toolbar, which also transitions to a "swap"-style icon when hovered. Lots of people think this means they can swap the image. The current swap icon to the right of that is barely recognized by anyone.

I think a text label is by far the best solution here. Worth noting that I've seen a handful of people expect to be able to double click as well, so I think that'd be a great enhancement here.

**Comment by kjellr** [comment]
> Part of the confusion is related to the block icon in the toolbar, which also transitions to a "swap"-style icon when hovered. Lots of people think this means they can swap the image. The current swap icon to the right of that is barely recognized by anyone.

That's a great point. When I designed the swap icon, I definitely didn't realize that. 😄 

<img width="239" alt="Screen Shot 2019-07-31 at 9 42 15 AM" src="https://user-images.githubusercontent.com/1202812/62216889-9c85b800-b377-11e9-937a-63689867c263.png">

**Comment by karmatosed** [comment]
My biggest concern around this is adding another link interface. For example Google does:

![image](https://user-images.githubusercontent.com/253067/62626058-679cd680-b91e-11e9-9660-558334a82da7.png)

@mapk what was the thinking in adding the link inline? Could we reuse the existing way links are added?

**Comment by mapk** [comment]
> @mapk what was the thinking in adding the link inline? Could we reuse the existing way links are added?

Good question, @karmatosed! So I thought about using the existing link interface, but then we'd have a popover on top of a popover which doesn't feel right. That's why I chose to put it inline. I'm sold on my solution though.

**Don't Do This**

![1 v2](https://user-images.githubusercontent.com/617986/62652949-58635c00-b911-11e9-832e-ec063d8e9abe.png)

I'm totaly open to a separate Dialog to inject a URL, but this also seems like a new interface option. Are we using a link interface that isn't a popover anywhere?

**Comment by kjellr** [comment]
> Are we using a link interface that isn't a popover anywhere?

Yeah, the Button block has one inline: 

![Screen Shot 2019-08-07 at 3 55 25 PM](https://user-images.githubusercontent.com/1202812/62653595-d59bca80-b92b-11e9-9a81-f288e0b3ec99.png)

This isn't really standard though, and other options are being explored in https://github.com/WordPress/gutenberg/issues/16495.

**Comment by mapk** [comment]
Using the Button block as an example, and merging it with my prior mockup, I came up with this.

![replace](https://user-images.githubusercontent.com/617986/62911601-cbb60500-bd39-11e9-8c9c-1b463ea8aa19.gif)

**Prototype:** 
https://www.figma.com/proto/dmv0ZfIjMxKr9YZGS5r9Gl/Replace-Image?node-id=32%3A1535&viewport=566%2C1244%2C1&scaling=min-zoom

**Comment by draganescu** [comment]
@mapk what happens when I search in that URL input, do I get suggestions on top of the dropdown?

**Comment by karmatosed** [comment]
The exploration fits for button block, however, some conversation was being had around unifying button block. Ultimately, I think looking holistically at link interactions is important. I am not sure it's important enough to hold up or block this issue. If that's case, then maybe we can open another ticket after to resolve this?

**Comment by mapk** [comment]
@senadir, I'd like to get some accessibility feedback here before committing to one solution. Ultimately, a PR can get started to work through some of the other pieces. I think the [last option here](https://github.com/WordPress/gutenberg/issues/11952#issuecomment-520667525) is the strongest right now.

cc @afercia @enriquesanchez @ryelle

**Comment by enriquesanchez** [comment]
I'm in favor of a text label instead of an icon, it's more obvious and accessible overall. As for the dropdown, we just need to make sure it's accessible (see #16473). I'd be more than happy to test once we have a PR.

**Comment by draganescu** [comment]
We have this PR I am working on here #16200 which will make possible to add this to any media component, not only image. 

Right now you can test using that or and inserting an audio block. It implemented the text button, the menu, media lib and file upload. 

Coming soon is the URL input.

**Comment by afercia** [comment]
Re: "Insert from URL":

- say I have this image URL: `https://cldup.com/cXyG__fTLN.jpg`
- shouldn't it work as embeds work? Just paste that in the editor and I should get an image block with the image already inserted
- instead, I get an "Embed" block
- the block tells me that kind of content can't be embedded (then, I'd argue it shouldn't try to embed an image in the first place):

<img width="765" alt="Screenshot 2019-08-23 at 14 15 09" src="https://user-images.githubusercontent.com/1682452/63592140-2a3c7800-c5b1-11e9-9add-7adf63f9ea6e.png">

- of course, for accessibility there should be some kind of UI but it could be very simplified
- also, when _replacing_ an existing image with a new one from a URL there should be some kind of UI (maybe also allow paste a URL over an existing image?)

**Comment by afercia** [comment]
Apart from the considerations above, the prototype from https://github.com/WordPress/gutenberg/issues/11952#issuecomment-520667525 looks promising to me from an a11y perspective. As long as it uses the dropdown menu, which already has good accessibility. The visible button text is an improvement for all users.

Question:
Will the following menu items open the media modal dialog?
- Media Library 
- Upload Image 
- Edit Image 

Also: worth reminding there's an open issue from the WPCampus report about _visible_ labelling only done via the placeholder attribute. To address this point, the URL field needs a visible `<label>` element. See https://core.trac.wordpress.org/ticket/47138
See also https://core.trac.wordpress.org/ticket/40331

**Comment by paaljoachim** [comment]
What about just a media text button?

![Media-link-toolbar](https://user-images.githubusercontent.com/5323259/63593309-062e6600-c5b4-11e9-917f-f762b1c2db14.jpg)

![Media-link-toolbar-no-icon](https://user-images.githubusercontent.com/5323259/63593316-09c1ed00-c5b4-11e9-93e6-19f3bc601dfb.jpg)

From the above it seems like the placeholder screen will be replaced by the toolbar replace image drop down. I do find it interesting as it could also incorporate the image link icon with a open image in new tab or no follow etc. Creating a element that can be used for all blocks that deal in someway with images. For audio/video etc the element could be somewhat adjusted.

![Media-toolbar-drop-down](https://user-images.githubusercontent.com/5323259/63593945-78ec1100-c5b5-11e9-9afc-0a86cc7138da.jpg)


![Media-toolbar-drop-down-video](https://user-images.githubusercontent.com/5323259/63593953-7c7f9800-c5b5-11e9-93e3-915fc4659c54.jpg)


![Media-toolbar-drop-down-audio](https://user-images.githubusercontent.com/5323259/63593959-7ee1f200-c5b5-11e9-9888-84a64a84a74e.jpg)

Link with the new tab etc V.
![Media-toolbar-drop-down-link](https://user-images.githubusercontent.com/5323259/63594274-4d1d5b00-c5b6-11e9-85da-f80f1f447d81.jpg)

**Comment by apeatling** [comment]
Coming here to +1 adding text labels. On WordPress.com we are regularly seeing users struggle to figure out how to replace images that already exist in a template. Looking through this thread, the clearest way to improve this IMO is to use labels like "Replace Image".
